R/verb-expand.R

```diff
@@ -0,0 +1,44 @@
+#' @importFrom tidyr expand
+#' @export
+expand.tbl_lazy <- function(data, ..., .name_repair = "check_unique") {
+  # TODO sort output?
+  # TODO wait for bugfix: distinct() ignores groups
+  # https://github.com/tidyverse/dbplyr/issues/535
+  dots <- quos_auto_name(quos(...))
+  dots <- purrr::discard(dots, rlang::quo_is_null)
+  names(dots) <- vctrs::vec_as_names(names(dots), repair = .name_repair)
+
+  if (is_empty(dots)) {
+    # TODO what should be returned here?
+    abort("No columns selected")
+  }
+
+  distinct_tables <- purrr::imap(
+    dots,
+    ~ {
+      # ugly hack to deal with `nesting()`
+      if (rlang::quo_is_call(.x) && rlang::call_name(.x) == "nesting") {
+        x_expr <- rlang::quo_get_expr(.x)
+        args <- rlang::call_args(x_expr)
+        distinct(data, !!!args)
+      } else {
+        distinct(data, !!.y := !!.x)
+      }
+    }
+  )
+
+  purrr::reduce(distinct_tables, left_join, by = character())
```

Comment:
Should this be a join or a union?

Reply:
This must be a cross join. This is the essential part to create all combinations between data. Basically the following `expand()`

    library(tidyr)
    library(dplyr)

    x <- c(1, 1, 2)
    y <- c("a", "b", "c")
    expand(
      tibble(x, y),
      x, y
    )
    #> # A tibble: 6 x 2
    #>       x y    
    #>   <dbl> <chr>
    #> 1     1 a    
    #> 2     1 b    
    #> 3     1 c    
    #> 4     2 a    
    #> 5     2 b    
    #> 6     2 c

translates to a cross join of two tibbles with unique values

    left_join(
      tibble(x) %>% distinct(),
      tibble(y) %>% distinct(),
      by = character()
    )
    #> # A tibble: 6 x 2
    #>       x y    
    #>   <dbl> <chr>
    #> 1     1 a    
    #> 2     1 b    
    #> 3     1 c    
    #> 4     2 a    
    #> 5     2 b    
    #> 6     2 c

<sup>Created on 2020-11-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
I think I put my comment in the wrong place — can `complete()` use a union instead of a join?

Reply:
No, this doesn't work. Union needs the same columns in both tables, so missing columns are created and filled with NULL. Then, the union doesn't work correctly anymore as you can see in the example

```r
# result with full join
#> # A tibble: 4 x 5
#>   group item_id item_name value1 value2
#>   <dbl>   <dbl> <chr>      <int>  <int>
#> 1     1       1 a              1      4
#> 2     1       2 b              3      6
#> 3     2       1 a             NA     NA
#> 4     2       2 b              2      5

# result with union
#> # A tibble: 7 x 5
#>   group item_id item_name value1 value2
#>   <dbl>   <dbl> <chr>      <int>  <int>
#> 1     1       1 a             NA     NA
#> 2     1       1 a              1      4
#> 3     1       2 b             NA     NA
#> 4     1       2 b              3      6
#> 5     2       1 a             NA     NA
#> 6     2       2 b             NA     NA
#> 7     2       2 b              2      5
```

Also this way the code is identical to `tidyr:::complete.data.frame()` up to a missing call to `reconstruct_tibble(data, full)`.

Reply:
Ok, makes sense.